scotty/pitch_angle_analysis.py

```diff
@@ -0,0 +1,858 @@
+        else:
+            x_start = bounds[0]
+            x_stop = bounds[1]
+            y_start = bounds[2]
+            y_stop = bounds[3]
+
+        x_array = np.linspace(x_start, x_stop, num=30)
+        y_array = np.linspace(y_start, y_stop, num=30)
+        Xgrid, Ygrid = np.meshgrid(x_array, y_array, indexing="ij")
+        Zgrid = spline_surface(x_array, y_array)
+
+        # Convert from poloidal flux to rho
+        if measure == "rho":
+            Zgrid = np.sqrt(Zgrid)
+
+        # Interpolate and apply the reflection mask to remove inaccessible regions
+        reflection_mask = (
+            self.dataset["reflection_mask"]
+            .loc[coords_dict]
+            .transpose(xdimension, ydimension)
+            .values
+        )
+        mask_function = RegularGridInterpolator(
+            (x_coordinates, y_coordinates),
+            reflection_mask,
+            method="nearest",
+            bounds_error=False,
+        )
+        interp_mask = mask_function((Xgrid, Ygrid))
+        Z_masked = np.ma.array(Zgrid, mask=interp_mask)
+
+        fig, ax = plt.subplots()
+        if mask_flag:
+            CS = ax.contour(
+                Xgrid, Ygrid, Z_masked, levels=20, cmap="plasma_r", corner_mask=True
+            )
+        else:
+            CS = ax.contour(
+                Xgrid, Ygrid, Zgrid, levels=20, cmap="plasma_r", corner_mask=True
+            )
+
+        fig.suptitle(f"Cutoff {title_label}, {const_angle_str}={const_angle}$^\circ$")
+        ax.set_xlabel("frequency/GHz")
+        ax.set_ylabel(f"{var_angle_str}/$^\circ$")
+        ax.clabel(CS, inline=True, fontsize=6)
+        fig.colorbar(CS)
+        if save_path:
+            plt.savefig(
+                save_path + f"cutoff_contour_{const_angle_str}{const_angle}.jpg",
+                dpi=200,
+            )
+        return (Xgrid, Ygrid, Zgrid, interp_mask)
+
+    # To be implemented
+    """ 
+    def plot_cutoff_angles(self, poloidal_angle):
+        return
+
+    def plot_opt_tor_vs_freq(self, poloidal_angle):
+        return
+    """
+
+
+class SweepFromNC(SweepDataset):
+    """Class for reading SweepDataset data written to a .nc file."""
+
+    def __init__(self, path_string):
+        with xr.open_dataset(path_string) as ds:
+            class_type = ds.attrs["class_type"]
+            if class_type != "SweepDataset":
+                raise ValueError(
+                    ".nc file does not have class_type = 'SweepDataset' attribute"
+                )
+            self.dataset = ds
+
+
+class PitchDiagnostic:
+    """Separate class for representing a particular pitch angle diagnostic design/setup.
+    Analysis methods take in Scotty simulation data in a SweepDataset class, and saves the
+    output in a dictionary of datasets. Each dataset corresponds to a type of analysis
+    output with variables corresponding to different SweepDataset equilibriums. Datasets
+    are organized into NetCDF groups when reading and writing to file.
```

Comment:
You might also look at [`Datatree`](https://github.com/xarray-contrib/datatree) to do this